internal/channel/plugin.go

```diff
@@ -1,19 +1,90 @@
 package channel
 
 import (
+	"bufio"
 	"encoding/json"
-	"errors"
 	"fmt"
-	"github.com/icinga/icinga-notifications/internal/contracts"
-	"github.com/icinga/icinga-notifications/internal/event"
-	"github.com/icinga/icinga-notifications/internal/recipient"
+	"github.com/icinga/icinga-notifications/internal/daemon"
 	"github.com/icinga/icinga-notifications/pkg/plugin"
 	"github.com/icinga/icinga-notifications/pkg/rpc"
-	"strings"
+	"github.com/icinga/icingadb/pkg/icingadb"
+	"github.com/icinga/icingadb/pkg/logging"
+	"go.uber.org/zap"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"time"
 )
 
-func (c *Channel) GetInfo() (*plugin.Info, error) {
-	result, err := c.rpcCall(plugin.MethodGetInfo, nil)
+type Plugin struct {
+	cmd    *exec.Cmd
+	rpc    *rpc.RPC
+	logger *zap.SugaredLogger
+
+	stopOnce sync.Once
+}
+
+// NewPlugin starts and returns a new plugin instance. If the start of the plugin fails, an error is returned
+func NewPlugin(pluginType string, logger *zap.SugaredLogger) (*Plugin, error) {
+	file := filepath.Join(daemon.Config().ChannelPluginDir, pluginType)
+	cmd := exec.Command(file)
+
+	writer, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stdin pipe: %w", err)
+	}
+
+	reader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stdout pipe: %w", err)
+	}
+
+	errPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stderr pipe: %w", err)
+	}
+
+	if err = cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start cmd: %w", err)
+	}
+
+	p := &Plugin{logger: logger.With(zap.Int("pid", cmd.Process.Pid))}
+
+	p.logger.Info("Successfully started channel plugin")
+
+	go forwardLogs(errPipe, p.logger)
+
+	p.cmd = cmd
+	p.rpc = rpc.NewRPC(writer, reader, p.logger)
```

Comment:
Now you can actually initialize all the fields right within the initializer. (`forwardLogs()` shouldn't mind if it's started after `NewRPC`, otherwise you could just do `logger = logger.With(...)` before instead of only setting it within the plugin).